packages/kbn-custom-integrations/src/state_machines/create/types.ts

```diff
@@ -0,0 +1,120 @@
+ * Side Public License, v 1.
+ */
+
+import { CustomIntegrationOptions, IntegrationError } from '../../types';
+import { CreateCustomIntegrationValue } from '../services/integrations_client';
+import { IndexedValidationErrors, ValidationErrors } from '../services/validation';
+
+export type CreateCustomIntegrationOptions = CustomIntegrationOptions;
+
+export interface WithTouchedFields {
+  touchedFields: Record<keyof CreateCustomIntegrationOptions, boolean>;
+}
+
+export type CreateInitialState = WithOptions & WithFields & WithPreviouslyCreatedIntegration;
+
+export interface WithOptions {
+  options: {
+    deletePrevious?: boolean;
+    resetOnCreation?: boolean;
+  };
+}
+
+export interface WithIntegrationName {
+  integrationName: CreateCustomIntegrationOptions['integrationName'];
+}
+
+export interface WithPreviouslyCreatedIntegration {
+  previouslyCreatedIntegration?: CreateCustomIntegrationOptions;
+}
+
+export interface WithDatasets {
+  datasets: CreateCustomIntegrationOptions['datasets'];
+}
+
+export interface WithFields {
+  fields: WithIntegrationName & WithDatasets;
+}
+
+export interface WithErrors {
+  errors: {
+    fields: Partial<{
+      integrationName: IntegrationError[];
+      datasets: IndexedValidationErrors;
+    }> | null;
+    general: IntegrationError | null;
+  };
+}
+
+export interface WithNullishErrors {
+  errors: null;
+}
+
+export type WithOptionalErrors = WithErrors | WithNullishErrors;
+
+export type DefaultCreateCustomIntegrationContext = WithOptions &
+  WithFields &
+  WithTouchedFields &
+  WithPreviouslyCreatedIntegration &
+  WithNullishErrors;
+
+export type CreateCustomIntegrationTypestate =
+  | {
+      value: 'uninitialized';
+      context: DefaultCreateCustomIntegrationContext;
+    }
+  | {
+      value: 'validating';
+      context: DefaultCreateCustomIntegrationContext & WithOptionalErrors;
+    }
+  | { value: 'valid'; context: DefaultCreateCustomIntegrationContext & WithNullishErrors }
+  | {
+      value: 'validationFailed';
+      context: DefaultCreateCustomIntegrationContext & WithErrors;
+    }
+  | { value: 'submitting'; context: DefaultCreateCustomIntegrationContext & WithNullishErrors }
+  | { value: 'success'; context: DefaultCreateCustomIntegrationContext & WithNullishErrors }
+  | { value: 'failure'; context: DefaultCreateCustomIntegrationContext & WithErrors }
+  | {
+      value: 'deletingPrevious';
+      context: DefaultCreateCustomIntegrationContext & WithNullishErrors;
+    };
+
+export type CreateCustomIntegrationContext = CreateCustomIntegrationTypestate['context'];
+
+export type CreateCustomIntegrationEvent =
+  | {
+      type: 'UPDATE_FIELDS';
+      fields: Partial<CreateCustomIntegrationOptions>;
+    }
+  | {
+      type: 'INITIALIZE';
+    }
+  | {
+      type: 'SAVE';
+    }
+  | {
+      type: 'RETRY';
+    }
+  // NOTE: These aren't ideal but they're more helpful than the DoneInvokeEvent<> and ErrorPlatformEvent types
```

Comment:
These types aren't ideal, but mixing `DoneInvokeEvent<>` and `ErrorPlatformEvent` is worse (in this case anyway). I'm definitely in favour of staying with the `onDone` and `onError` callbacks as promises fulfil these scenarios best (callbacks can't use async / await etc).

If this becomes unwieldy we could potentially have a convention where the top-level key on `data` actually acts as a type in itself, so that TypeScript can always narrow down to something useful.

E.g. `data.submittingResponse.<CreateCustomIntegrationValue>`, then if the actions check for `data.submittingResponse` the type becomes sufficiently narrow.